### PR TITLE
[Controllers Branch] - various updates

### PR DESCRIPTION
**Pull request by ZappoMan**
* merges latest upstream with controllers
* applies Huffman's recommended improvements to Sam's recent PR
* change the "join()" filter to be "makeAxis()"
* rename hardware devices to not have device ID in the name
* don't include the Standard device in Controllers.Hardware
* thread safety for JS based filters

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 75c59fad992f7e8d7d8bbd8f187620824a68a8cf available at: https://gist.github.com/2799aa92d6d14589c664<br/><br/>No installer was built for this commit

**Comment by hifi-gustavo** [comment]
Build console output for commit: f90f0ed1f342c37b2e8187981f10022a8021a425 available at: https://gist.github.com/2799aa92d6d14589c664<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6105.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6105.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: e888391e7edb9e5c17ad5d9aca65c5dbbd8dda3c available at: https://gist.github.com/2799aa92d6d14589c664<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6105.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6105.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 293b7c12e151b49305b4d87137816aee387c4e1b available at: https://gist.github.com/2799aa92d6d14589c664<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6105.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6105.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 251a55b1fb255355ec5f8fbb37b96f1f9c982018 available at: https://gist.github.com/2799aa92d6d14589c664<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6105.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6105.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7d48fe918729f9cd6f5d2014a3cb7673b2a0becb available at: https://gist.github.com/2799aa92d6d14589c664<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6105.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6105.dmg

File: examples/controllers/controllerMappings.js [hunk 1]
```diff
@@ -70,7 +70,7 @@ mapping2.enable();
 
 Controller.enableMapping("example2");
 */
-var mapping3 = Controller.loadMapping("E:/Github/hifi/examples/controllers/example3.json");
+var mapping3 = Controller.loadMapping(Script.resolvePath("example3.json"));
```
Comment (ZappoMan) [inline examples/controllers/controllerMappings.js]: Huffman's recommendation on Sam's PR

File: examples/html/entityProperties.html [hunk 2]
```diff
@@ -1035,9 +1035,9 @@
         <div class="property">
             <div class="label">Position</div>
             <div class="value">
-                <div class="input-area">X <br><input class="coord" type='number' id="property-pos-x"></div>
-                <div class="input-area">Y <br><input class="coord" type='number' id="property-pos-y"></div>
-                <div class="input-area">Z <br><input class="coord" type='number' id="property-pos-z"></div>
+                <div class="input-area ">X<input class="coord" type='number' id="property-pos-x"><div class="prop-x"></div></div>
```
Comment (ZappoMan) [inline examples/html/entityProperties.html]: this is just a merge from master

File: examples/html/style.css [hunk 3]
```diff
@@ -342,3 +343,21 @@ input#property-name {
 	margin: 5px;
 	border: 1px solid white;
 }
+
+.prop-x {
```
Comment (ZappoMan) [inline examples/html/style.css]: this is just a merge from master

File: libraries/controllers/src/controllers/ScriptingInterface.cpp [hunk 4]
```diff
@@ -220,12 +233,13 @@ namespace controller {
...
             if (request->getResult() == ResourceRequest::Success) {
-                return parseMapping(QString(request->getData()));
+                result = parseMapping(QString(request->getData()));
             } else {
                 qDebug() << "Failed to load mapping url <" << jsonUrl << ">" << endl;
-                return nullptr;
             }
+            request->deleteLater();
         }
+        return result;
```
Comment (ZappoMan) [inline libraries/controllers/src/controllers/ScriptingInterface.cpp]: fixes some warnings that not all paths returned a value

File: libraries/controllers/src/controllers/ScriptingInterface.cpp [hunk 5]
```diff
@@ -220,12 +233,13 @@ namespace controller {
             }
 
             if (request->getResult() == ResourceRequest::Success) {
-                return parseMapping(QString(request->getData()));
+                result = parseMapping(QString(request->getData()));
             } else {
                 qDebug() << "Failed to load mapping url <" << jsonUrl << ">" << endl;
-                return nullptr;
             }
+            request->deleteLater();
```
Comment (ZappoMan) [inline libraries/controllers/src/controllers/ScriptingInterface.cpp]: Huffman's CR suggestion from Sam's PR

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7956d737ab1b50fecf2c72253ff7a612e4796975 available at: https://gist.github.com/2799aa92d6d14589c664<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6105.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6105.dmg

File: libraries/controllers/src/controllers/ScriptingInterface.cpp [hunk 6]
```diff
@@ -60,24 +61,35 @@ namespace controller {
...
-        }
+    float ScriptEndpoint::value() {
+        updateValue();
+        return _lastValue;
+    }
 
-        virtual float value() {
-            float result = (float)_callable.call().toNumber();
-            return result;
+    void ScriptEndpoint::updateValue() {
```
Comment (ZappoMan) [inline libraries/controllers/src/controllers/ScriptingInterface.cpp]: this work adds thread safety to JS functions in filter mappings

File: libraries/controllers/src/controllers/ScriptingInterface.cpp [hunk 7]
```diff
@@ -479,31 +493,34 @@ namespace controller {
...
-            // Expose the IDs to JS
-            _hardware.insert(deviceName, createDeviceMap(device));
-
-            // Create the endpoints
-            for (const auto& inputMapping : device->getAvailabeInputs()) {
-                const auto& input = inputMapping.first;
-                // Ignore aliases
-                if (_endpoints.count(input)) {
+            auto deviceID = deviceMapping.first;
+            if (deviceID != userInputMapper->getStandardDeviceID()) {
```
Comment (ZappoMan) [inline libraries/controllers/src/controllers/ScriptingInterface.cpp]: this change removes the "Standard" device from the "Controllers.Hardware"

File: libraries/controllers/src/controllers/ScriptingInterface.h [hunk 8]
```diff
@@ -138,6 +139,25 @@ namespace controller {
         MappingMap _mappingsByName;
         MappingStack _activeMappings;
     };
+
+    class ScriptEndpoint : public Endpoint {
```
Comment (ZappoMan) [inline libraries/controllers/src/controllers/ScriptingInterface.h]: for MOC to work the class definition needs to be int he header - this is for scripting thread safety

File: libraries/controllers/src/controllers/UserInputMapper.cpp [hunk 9]
```diff
@@ -31,14 +31,28 @@ UserInputMapper::UserInputMapper() {
...
+        _deviceCounts[deviceName] = 0;
+    }
+    _deviceCounts[deviceName] += 1;
+    return _deviceCounts[deviceName];
+}
 
-bool UserInputMapper::registerDevice(uint16 deviceID, const DeviceProxy::Pointer& proxy){
-    proxy->_name += " (" + QString::number(deviceID) + ")";
-    _registeredDevices[deviceID] = proxy;
+bool UserInputMapper::registerDevice(uint16 deviceID, const DeviceProxy::Pointer& proxy) {
```
Comment (ZappoMan) [inline libraries/controllers/src/controllers/UserInputMapper.cpp]: this work fixes the naming for input devices to not have the (random) device ID included

File: libraries/controllers/src/controllers/impl/MappingBuilderProxy.cpp [hunk 10]
```diff
@@ -39,7 +39,7 @@ QObject* MappingBuilderProxy::from(const Endpoint::Pointer& source) {
     return new RouteBuilderProxy(_parent, _mapping, route);
 }
 
-QObject* MappingBuilderProxy::join(const QJSValue& source1, const QJSValue& source2) {
+QObject* MappingBuilderProxy::makeAxis(const QJSValue& source1, const QJSValue& source2) {
```
Comment (ZappoMan) [inline libraries/controllers/src/controllers/impl/MappingBuilderProxy.cpp]: join() filter is now makeAxis()

File: libraries/entities/src/ParticleEffectEntityItem.cpp [hunk 11]
```diff
@@ -644,14 +644,13 @@ void ParticleEffectEntityItem::stepSimulation(float deltaTime) {
 
     // update particles between head and tail
     for (quint32 i = _particleHeadIndex; i != _particleTailIndex; i = (i + 1) % _maxParticles) {
-        _particleLifetimes[i] -= deltaTime;
+        _particleLifetimes[i] += deltaTime;
```
Comment (ZappoMan) [inline libraries/entities/src/ParticleEffectEntityItem.cpp]: these changes are a merge from master

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8578b40236d9f4dfbd9b4feb0098133ce94a2ece available at: https://gist.github.com/2799aa92d6d14589c664<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6105.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6105.dmg

File: libraries/controllers/src/controllers/ScriptingInterface.h (hunk 8 above)
Comment (jherico) [inline libraries/controllers/src/controllers/ScriptingInterface.h]: actually you can explicitly declare `#include "ScriptingInterface.moc"` in the cpp file and this doesn't need to be in the header, but no big deal.

Comment (ZappoMan) [inline libraries/controllers/src/controllers/ScriptingInterface.h]: but if you take that approach, the build will fail on a fresh checkout until the MOC generation has been run at least once.

**Pull request merged**